### ISSUE TITLE
bson::decoder::error is private.

### ISSUE THREAD

**Issue by ruseinov** [mongodb/bson-rust#151]
Which makes it impossible to implement an automatic conversion using ```impl From .. ```

**Comment by saghm** [comment]
You can import it from the top-level of the crate, i.e. `bson::DecoderError`.

**Comment by ruseinov** [comment]
My bad, closing this.

**Comment by ruseinov** [comment]
@saghm btw the rust driver is missing index operations as I'm sure you know. Excuse me for bringing it up here, but the rust driver does not allow public to create issues. In case you want/need something to contribute this - I'd be happy to assist.

**Comment by saghm** [comment]
We track issues for the Rust driver on JIRA; see [the instructions listed in the README](https://github.com/mongodb/mongo-rust-driver#bug-reporting--feature-requests) if you're having trouble with creating an issue.

As for the index operations, we currently haven't prioritized implementing them based on the fact that most users will presumably only need to perform them once when initially creating the application and usually do so in the shell. Is there something specific you need index operations for in the driver itself, rather than creating/checking indexes on an ad-hoc basis using the shell?

**Comment by ruseinov** [comment]
I'm working a Rust application that uses MongoDB as a database at the moment and I'm forced to use `docker-entrypoint-initdb.d` for development to add indices. It would be way more convenient to just ensure all this indices on application start as the application already knows how the models look. Having to use 3rd party scrips for CI and docker dev setups. 

I could take the implementation on, it's going to take a while, but if this is not a priority - should be alright. I just didn't want to implement something that's already being worked on by somebody else. 

I saw that you guys merged the async changes yesterday, which is amazing, as I've been waiting for that too.

**Comment by ruseinov** [comment]
@saghm ^^

**Comment by saghm** [comment]
In general, we consider creating indexes as part of the application instead of separately as a bit of anti-pattern. I understand that your use case requires it though; for now, I'd suggest using the [Database::run_command](https://docs.rs/mongodb/0.9.2/mongodb/struct.Database.html#method.run_command) helper to manually run the [`createIndexes` command](https://docs.mongodb.com/manual/reference/command/createIndexes/).

**Comment by ruseinov** [comment]
@saghm That works. It depends on the application I guess, when you want to keep it simple stupid with one deliverable backend artifact - it seems only natural to do this within the app in order to avoid extra burden. 

I'm also looking forward to transactions support for some edge-cases in the app, will look into it to see if I might be able to contribute some of that work or at least prototype something to use for now.

**Comment by saghm** [comment]
Transactions are one of the first things we plan on working on after we stabilize the API (the last major piece of which is finishing up the async work). There's a bit of prerequisite work that will need to be done first to enable transactions though.

**Comment by ruseinov** [comment]
@saghm Sounds good! Let me know if I can help in any way.
